### PR TITLE
[Optimization] Optimize pickling of heterograph

### PR DESCRIPTION
**Pull request by soodoshll**
## Description

Implement the major part in C to avoid calling too many ``__getstate__`` functions in python.

**Time for transmitting 1000 graphs(from CoraDataset)**
- old: 2.20s
- new: 1.03s 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by BarclayII** [comment]
When you're finished please request a review.  Thanks for your effort!

File: src/graph/pickle.cc [hunk 1]
```diff
@@ -14,85 +17,136 @@ namespace dgl {
...
+  strm->Write(fmts);
+  strm->Write(flags);
   return states;
 }
 
 HeteroGraphPtr HeteroUnpickle(const HeteroPickleStates& states) {
-  const auto metagraph = states.metagraph;
-  const auto &num_nodes_per_type = states.num_nodes_per_type;
-  CHECK_EQ(states.adjs.size(), metagraph->NumEdges());
+  char *buf = const_cast<char *>(states.meta.c_str());  // a readonly stream?
```
Comment (VoVAllen) [inline src/graph/pickle.cc:58]: Why not still using `MemoryStringStream`?

File: src/graph/pickle.cc [hunk 2]
```diff
@@ -14,85 +17,136 @@ namespace dgl {
...
   HeteroPickleStates states;
-  states.metagraph = graph->meta_graph();
-  states.num_nodes_per_type = graph->NumVerticesPerType();
-  states.adjs.resize(graph->NumEdgeTypes());
+  dmlc::MemoryStringStream ofs(&states.meta);
+  dmlc::Stream *strm = &ofs;
+  strm->Write(ImmutableGraph::ToImmutable(graph->meta_graph()));
+  strm->Write(graph->NumVerticesPerType());
+  std::vector<SparseFormat> fmts;
+  std::vector<char> flags;
```
Comment (VoVAllen) [inline src/graph/pickle.cc:25]: Why does `vector<char>` needed here? You can read/write booleans separately, why you prefer first constructing a vector and then write them?

**Comment by VoVAllen** [comment]
I'm a bit concerned about the way of implementation. This would only accelerate pytorch case, right? And according to pytorch, this acceleration should only be applicable for the multiprocessing scenario, which should be done by `ForkingPickler.register(DGLHeteroGraph, reduce_function)` instead of changing `setstate` and `getstate` for the original codes. @jermainewang 

My expectation would be only register this pickling method when using pytorch under multiprocessing scenario. Otherwise a naive implementation(or dgl's shared memory tensor in the future) could be fine.

File: src/graph/pickle.cc (hunk 1 above)
Comment (soodoshll) [reply src/graph/pickle.cc:58]: This line is to drop the `const`. It is still necessary if using `MemoryStringStream`?

Comment (VoVAllen) [reply src/graph/pickle.cc:58]: You can directly use `dmlc::MemoryStringStream ifs(&states.meta);` as how you write.

**Comment by jermainewang** [comment]
> I'm a bit concerned about the way of implementation. This would only accelerate pytorch case, right? And according to pytorch, this acceleration should only be applicable for the multiprocessing scenario, which should be done by `ForkingPickler.register(DGLHeteroGraph, reduce_function)` instead of changing `setstate` and `getstate` for the original codes. @jermainewang

What's the concern here? Do you foresee this implementation will slowdown pickle/unpickle in frameworks other than PyTorch?
 
> My expectation would be only register this pickling method when using pytorch under multiprocessing scenario. Otherwise a naive implementation(or dgl's shared memory tensor in the future) could be fine.

What do you mean by a naive implementation? Is the the current implementation or saving everything in one buffer?

**Comment by VoVAllen** [comment]
My only concern is whether the current implementation is stable(whether there would be compatibility issue in the future). Pickle to file is different from pickle in multiprocessing, since there's no compatibility issue in the multiprocessing scenario. Therefore the purpose of multiprocessing pickle could be fast, but the purpose of file pickle could be stable.

For example, maybe we will need to turn to other methods to optimize the multiprocessing performance in Tensorflow, which might break the compatibility of file pickle in the future. Using `ForkingPickler` could be address this kind of issue.

File: src/graph/pickle.cc [hunk 3]
```diff
@@ -107,8 +186,23 @@ DGL_REGISTER_GLOBAL("heterograph_index._CAPI_DGLHeteroPickle")
 DGL_REGISTER_GLOBAL("heterograph_index._CAPI_DGLHeteroUnpickle")
 .set_body([] (DGLArgs args, DGLRetValue* rv) {
     HeteroPickleStatesRef ref = args[0];
-    HeteroGraphPtr graph = HeteroUnpickle(*ref.sptr());
+    HeteroGraphPtr graph = ref->version ?
```
Comment (VoVAllen) [inline src/graph/pickle.cc:189]: Please use switch/if statement and check all the condition. (raise error by LOG(FATAL) if ref->version is not 0 or 1)

**Comment by VoVAllen** [comment]
Thanks!

**Pull request merged**